### DISCUSSION TITLE
Feature Request: Support FTC532 Touch Controller

### DISCUSSION THREAD

**Discussion by vic42** [arendst/Tasmota#10185] [Ideas]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._
Chinese touch switches utilizing the FTC532 touch chip are not supported

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [X] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [X] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [X] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [X] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [X] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [X] Device used (e.g., Sonoff Basic): Teekar Smart Curtain Switch
- [X] Tasmota binary firmware version number used: 9.1.0.2
  - [ ] Pre-compiled
  - [X] Self-compiled
- [ ] Flashing tools used: _____
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:

N/A
```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:

N/A
```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:

N/A
```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:

N/A
```

### TO REPRODUCE
_Steps to reproduce the behavior:_
Flash the switch with Tasmota

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

The three touch buttons should be detectable as switch inputs to Tasmota

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

![image](https://user-images.githubusercontent.com/3363362/102694014-45a3ab80-4216-11eb-9c40-f1656526ae89.png)

LA traces drom top to bottom:
1. no key pressed
2. left key pressed
3. middle key pressed
4. right key pressed
5. all three keys pressed

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

There is just one line (Pin4 of 8) of the FTC532 chip connected to the ESP module through GPIO14. I analyzed the transmission protocol. The FTC532 is constantly sending the touch pads' status.

I assume the protocol is capable of reporting 2 groups of 4 touch sensors each.
Here is my protocol decode:

```
Frame Period: 21.85 ms (58 cycles)
Bit Period: 377 µs

IDLE  = 3015 µs
LEAD  =  755 µs
SHORT =  377 µs
LONG  = 1130 µs

T = 377 µs

IDLE  = 8T HIGH
LEAD  = 2T LOW, 1T HIGH
SHORT = 1T LOW, 1T HIGH
LONG  = 3T LOW, 1T HIGH

LEAD-IN = 3015 µs HIGH, 755 µs LOW

S = 377 µs HIGH,  377 µs LOW
L = 377 µs HIGH, 1130 µs LOW

          GROUP1   GROUP2
---------------------------
ALL OFF: SSSSLLLL SSSSLLLL
 1 ON  : LSSSSLLL SSSSLLLL
 2 ON  : SLSSLSLL SSSSLLLL
 3 ON  : SSLSLLSL SSSSLLLL
 4 ON  : SSSLLLLS SSSSLLLL
 5 ON  : SSSSLLLL LSSSSLLL
 6 ON  : SSSSLLLL SLSSLSLL
 7 ON  : SSSSLLLL SSLSLLSL
 8 ON  : SSSSLLLL SSSLLLLS
 
 
1-3 ON : LLLSSSSL SSSSLLLL
ALL ON : LLLLSSSS LLLLSSSS

Example timing ALL OFF frame in clock cycles T=377µs, triggering on rising edge:
IDLE-22222444422224444-IDLE

```

I know programming, yet I'm looking for engagement of Devs more experienced with the Tasmota code.
Opened this issue on behalf of the original requestant to keep it on the "radar".
**(Please, remember to close the issue when the problem has been addressed)**

**Comment by snookyly** [comment]
As the original requestant I would like to thank @vic42 for his great analysis. The FTC532 touch controller is assembled to the 4th gen Teekar curtain switch (https://www.amazon.de/dp/B07XJWYZR4) and the Teekar 2-gang wall switch (https://www.amazon.de/dp/B07SLJ8JYJ). I'd would be great if both devices get the tasmota support.

**Reply by vic42** [reply]
I made some progress in creating a driver.
![Spectacle x32440](https://user-images.githubusercontent.com/3363362/102795073-318db480-43a4-11eb-9f4a-003436003371.png)
It appears a bit slow in the log but that may be a logging issue.

**Comment by vic42** [comment]
Driver programming finished.
The driver is in final testing and proof-reading stage now.
There will be a PR for it soon.

**Comment by vic42** [comment]
Driver has been merged into Tasmota Development branch.
`#define USE_FTC532` to include it in your build.

**Comment by snookyly** [comment]
Successfully tested. Thanks for implementing.
Can be closed.
